File: backend/app/api/graph.py
"""
Graph-related API routes
Uses project context mechanism with server-side persistent state
"""

import os
import traceback
import threading
from flask import request, jsonify

from . import graph_bp
from ..config import Config
from ..services.ontology_generator import OntologyGenerator
from ..services.graph_builder import GraphBuilderService
from ..services.text_processor import TextProcessor
from ..utils.file_parser import FileParser
from ..utils.logger import get_logger
from ..models.task import TaskManager, TaskStatus
from ..models.project import ProjectManager, ProjectStatus

# Get logger
logger = get_logger('pubop.api')


def allowed_file(filename: str) -> bool:
    """Check if file extension is allowed"""
    if not filename or '.' not in filename:
        return False
    ext = os.path.splitext(filename)[1].lower().lstrip('.')
    return ext in Config.ALLOWED_EXTENSIONS


# ============== Project Management API ==============

@graph_bp.route('/project/<project_id>', methods=['GET'])
def get_project(project_id: str):
    """
    Get project details
    """
    project = ProjectManager.get_project(project_id)
    
    if not project:
        return jsonify({
            "success": False,
            "error": f"Project does not exist: {project_id}"
        }), 404
    
    return jsonify({
        "success": True,
        "data": project.to_dict()
    })


@graph_bp.route('/project/<project_id>', methods=['DELETE'])
def delete_project(project_id: str):
    """
    Delete project
    """
    success = ProjectManager.delete_project(project_id)
    
    if not success:
        return jsonify({
            "success": False,
            "error": f"Project does not exist: {project_id}"
        }), 404
    
    return jsonify({
        "success": True,
        "message": "Project deleted successfully"
    })


@graph_bp.route('/projects', methods=['GET'])
def list_projects():
    """
    Get list of all projects
    """
    projects = ProjectManager.list_projects()
    
    return jsonify({
        "success": True,
        "data": {
            "projects": [p.to_dict() for p in projects],
            "total": len(projects)
        }
    })


# ============== API 1: Generate Ontology ==============

@graph_bp.route('/ontology/generate', methods=['POST'])
def generate_ontology():
    """
   API 1: Upload files and generate ontology definition
    
    Request (multipart/form-data):
        - files: One or more files (PDF, MD, TXT)
        - simulation_requirement: Simulation requirement description (required)
        - additional_context: Additional context (optional)
        - project_name: Project name (optional, default "Unnamed Project")
    
    Returns:
        {
            "success": true,
            "data": {
                "project_id": "proj_xxxx",
                "ontology": {
                    "entity_types": [...],
                    "edge_types": [...]
                },
                "analysis_summary": "...",
                "total_text_length": 12345
            }
        }
    
    Workflow:
        1. Check configuration
        2. Parse request parameters
        3. Create project
        4. Extract text from uploaded files
        5. Call LLM to generate ontology
        6. Save ontology to project
        7. Return ontology definition
    """
    try:
        logger.info("=== Starting ontology generation ===")
        
        # Debug: Log request details
        logger.info(f"Request content type: {request.content_type}")
        logger.info(f"Request content length: {request.content_length}")
        logger.info(f"Request files keys: {list(request.files.keys())}")
        logger.info(f"Request form keys: {list(request.form.keys())}")
        
        # Get form data
        files = request.files.getlist('files')
        simulation_requirement = request.form.get('simulation_requirement', '').strip()
        additional_context = request.form.get('additional_context', '').strip()
        project_name = request.form.get('project_name', 'Unnamed Project').strip()
        
        logger.info(f"Project name: {project_name}")
        logger.info(f"Simulation requirement length: {len(simulation_requirement)}")
        logger.info(f"Files received: {len(files)}")
        
        # Log each file info
        for i, f in enumerate(files):
            logger.info(f"File {i}: filename='{f.filename}', content_type='{f.content_type}'")
        
        # Validate parameters
        if not files:
            logger.error("No files in request")
            return jsonify({
                "success": False,
                "error": "At least one file required"
            }), 400
        
        if not simulation_requirement:
            logger.error("No simulation requirement provided")
            return jsonify({
                "success": False,
                "error": "Simulation requirement is required"
            }), 400
        
        # Create project
        project = ProjectManager.create_project(name=project_name)
        project.simulation_requirement = simulation_requirement
        project.additional_context = additional_context if additional_context else None
        ProjectManager.save_project(project)
        logger.info(f"Created project: {project.project_id}")
        
        # Extract file text
        document_texts = []
        for idx, file in enumerate(files):
            logger.info(f"Processing file {idx}: filename='{file.filename}'")
            
            if not file or not file.filename:
                logger.warning(f"Skipping file {idx}: empty file or filename")
                continue
            
            if not allowed_file(file.filename):
                logger.error(f"File type not supported: {file.filename}")
                return jsonify({
                    "success": False,
                    "error": f"File type not supported: {file.filename}"
                }), 400
            
            try:
                # Save uploaded file temporarily
                temp_dir = Config.UPLOAD_FOLDER
                os.makedirs(temp_dir, exist_ok=True)
                temp_path = os.path.join(temp_dir, file.filename)
                
                # Reset file stream position and read bytes to avoid corruption
                file.stream.seek(0)
                file_bytes = file.stream.read()
                
                logger.info(f"File {file.filename}: read {len(file_bytes)} bytes from stream")
                
                if len(file_bytes) == 0:
                    logger.error(f"File {file.filename} has 0 bytes - stream may be corrupted")
                    return jsonify({
                        "success": False,
                        "error": f"File {file.filename} is empty or corrupted during upload"
                    }), 400
                
                # Write bytes directly to file
                with open(temp_path, 'wb') as f:
                    f.write(file_bytes)
                
                logger.info(f"Saved file {file.filename} to {temp_path}")
                
                # Extract text using FileParser
                text = FileParser.extract_text(temp_path)
                
                # Clean up temp file
                os.remove(temp_path)
                
                if text:
                    document_texts.append(text)
                    logger.info(f"Extracted {len(text)} characters from {file.filename}")
                else:
                    logger.warning(f"No text extracted from {file.filename}")
            except Exception as e:
                logger.error(f"Failed to extract text from {file.filename}: {e}")
                logger.error(f"Traceback: {traceback.format_exc()}")
                return jsonify({
                    "success": False,
                    "error": f"Failed to parse file {file.filename}: {str(e)}"
                }), 500
        
        if not document_texts:
            logger.error("No valid text extracted from any files")
            return jsonify({
                "success": False,
                "error": "No valid text extracted from uploaded files"
            }), 400
        
        # Save extracted text
        all_text = "\n\n".join(document_texts)
        project.total_text_length = len(all_text)
        ProjectManager.save_extracted_text(project.project_id, all_text)
        logger.info(f"Text extraction complete, total {len(all_text)} characters")
        
        # Generate ontology
        logger.info("Calling LLM to generate ontology definition...")
        try:
            generator = OntologyGenerator()
            ontology = generator.generate(
                document_texts=document_texts,
                simulation_requirement=simulation_requirement,
                additional_context=additional_context if additional_context else None
            )
        except Exception as e:
            logger.error(f"Ontology generation failed: {type(e).__name__}: {str(e)}")
            logger.error(f"Traceback: {traceback.format_exc()}")
            raise
        
        # Save ontology to project
        entity_count = len(ontology.get("entity_types", []))
        edge_count = len(ontology.get("edge_types", []))
        logger.info(f"Ontology generation complete: {entity_count} entity types, {edge_count} relationship types")
        
        project.ontology = {
            "entity_types": ontology.get("entity_types", []),
            "edge_types": ontology.get("edge_types", [])
        }
        project.analysis_summary = ontology.get("analysis_summary", "")
        project.status = ProjectStatus.ONTOLOGY_GENERATED
        ProjectManager.save_project(project)
        logger.info(f"=== Ontology generation complete === Project ID: {project.project_id}")
        
        return jsonify({
            "success": True,
            "data": {
                "project_id": project.project_id,
                "ontology": project.ontology,
                "analysis_summary": project.analysis_summary,
                "total_text_length": project.total_text_length
            }
        })
        
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e),
            "traceback": traceback.format_exc()
        }), 500


# ============== API 2: Build Graph ==============

@graph_bp.route('/build', methods=['POST'])
def build_graph():
    """
    API 2: Build graph based on project_id
    
    Request (JSON):
        {
            "project_id": "proj_xxxx",  // Required, from API 1
            "graph_name": "Graph name",    // Optional
            "chunk_size": 500,          // Optional, default 500
            "chunk_overlap": 50         // Optional, default 50
        }
    
    Returns:
        {
            "success": true,
            "data": {
                "task_id": "task_xxxx"
            }
        }
    
    Workflow:
        1. Check configuration
        2. Validate project
        3. Get ontology from project
        4. Get text from project
        5. Create background task to build graph
        6. Return task_id (use /task/<task_id> to query progress)
    """
    try:
        logger.info("=== Starting graph build ===")
        
        # Check configuration
        errors = []
        if not Config.NEO4J_URI or not Config.NEO4J_PASSWORD:
            errors.append("Neo4j connection not configured (NEO4J_URI, NEO4J_PASSWORD)")
        if errors:
            logger.error(f"Configuration error: {errors}")
            return jsonify({
                "success": False,
                "error": "Configuration error: " + "; ".join(errors)
            }), 500
        
        # Parse request
        data = request.get_json()
        project_id = data.get('project_id', '').strip()
        graph_name = data.get('graph_name', '').strip() or "Fishi Graph"
        chunk_size = int(data.get('chunk_size', Config.DEFAULT_CHUNK_SIZE))
        chunk_overlap = int(data.get('chunk_overlap', Config.DEFAULT_CHUNK_OVERLAP))
        
        logger.debug(f"Project ID: {project_id}")
        logger.debug(f"Graph name: {graph_name}")
        logger.debug(f"Chunk size: {chunk_size}, overlap: {chunk_overlap}")
        
        # Validate parameters
        if not project_id:
            return jsonify({
                "success": False,
                "error": "project_id is required"
            }), 400
        
        # Get project
        project = ProjectManager.get_project(project_id)
        if not project:
            return jsonify({
                "success": False,
                "error": f"Project does not exist: {project_id}"
            }), 404
        
        # Check if ontology exists
        if not project.ontology or project.status != ProjectStatus.ONTOLOGY_GENERATED:
            return jsonify({
                "success": False,
                "error": "Project ontology not generated, please call /ontology/generate first"
            }), 400
        
        # Get extracted text
        text = ProjectManager.get_extracted_text(project_id)
        if not text:
            return jsonify({
                "success": False,
                "error": "Project text not found, file may have been deleted"
            }), 404
        
        logger.info(f"Text length: {len(text)} characters")
        
        # Create task manager
        task_manager = TaskManager()
        
        # Start graph build in background thread
        def build_graph_worker(task_id: str):
            try:
                # Create graph builder service
                builder = GraphBuilderService()
                
                # Split text into chunks
                task_manager.update_task(
                    task_id,
                    progress=5,
                    message="Splitting text into chunks..."
                )
                chunks = TextProcessor.split_text(text, chunk_size, chunk_overlap)
                logger.info(f"Text split into {len(chunks)} chunks")
                
                # Build graph asynchronously
                task_manager.update_task(
                    task_id,
                    progress=10,
                    message="Starting graph build..."
                )
                
                # Call graph builder's internal async method
                # The graph builder will update progress through task_manager
                graph_task_id = builder.build_graph_async(
                    text=text,
                    ontology=project.ontology,
                    graph_name=f"{project.name} - {graph_name}",
                    chunk_size=chunk_size,
                    chunk_overlap=chunk_overlap
                )
                
                # Wait for graph build to complete and forward progress
                while True:
                    graph_task = task_manager.get_task(graph_task_id)
                    if not graph_task:
                        break
                    
                    # Forward progress
                    task_manager.update_task(
                        task_id,
                        status=graph_task.status,
                        progress=graph_task.progress,
                        message=graph_task.message
                    )
                    
                    if graph_task.status in [TaskStatus.COMPLETED, TaskStatus.FAILED]:
                        if graph_task.status == TaskStatus.COMPLETED:
                            # Update project with graph_id
                            project.graph_id = graph_task.result.get("graph_id")
                            project.status = ProjectStatus.GRAPH_BUILT
                            ProjectManager.save_project(project)
                            
                            task_manager.complete_task(
                                task_id,
                                result=graph_task.result
                            )
                        else:
                            task_manager.fail_task(task_id, graph_task.error)
                        break
                    
                    import time
                    time.sleep(1)
                
            except Exception as e:
                logger.error(f"Graph build failed: {str(e)}\n{traceback.format_exc()}")
                task_manager.fail_task(task_id, str(e))
        
        # Create task
        task_id = task_manager.create_task(
            task_type="graph_build",
            metadata={
                "project_id": project_id,
                "graph_name": graph_name,
                "chunk_size": chunk_size,
                "chunk_overlap": chunk_overlap
            }
        )
        
        # Start background thread
        thread = threading.Thread(target=build_graph_worker, args=(task_id,))
        thread.daemon = True
        thread.start()
        
        logger.info(f"Graph build task created: {task_id}")
        
        return jsonify({
            "success": True,
            "data": {
                "task_id": task_id
            }
        })
        
    except Exception as e:
        logger.error(f"Graph build request failed: {str(e)}\n{traceback.format_exc()}")
        return jsonify({
            "success": False,
            "error": str(e),
            "traceback": traceback.format_exc()
        }), 500


# ============== Task Status Query ==============

@graph_bp.route('/task/<task_id>', methods=['GET'])
def get_task_status(task_id: str):
    """
    Query task status and progress
    
    Returns:
        {
            "success": true,
            "data": {
                "task_id": "task_xxxx",
                "status": "processing|completed|failed",
                "progress": 50,  // 0-100
                "message": "Current status message",
                "result": {},    // Only when completed
                "error": ""      // Only when failed
            }
        }
    """
    task_manager = TaskManager()
    task = task_manager.get_task(task_id)
    
    if not task:
        return jsonify({
            "success": False,
            "error": f"Task does not exist: {task_id}"
        }), 404
    
    return jsonify({
        "success": True,
        "data": task.to_dict()
    })


# ============== Get Graph Data ==============

@graph_bp.route('/data/<graph_id>', methods=['GET'])
def get_graph_data(graph_id: str):
    """
    Get graph data (nodes and edges)
    """
    try:
        if not Config.NEO4J_URI:
            return jsonify({
                "success": False,
                "error": "Neo4j connection not configured"
            }), 500
        
        builder = GraphBuilderService()
        graph_data = builder.get_graph_data(graph_id)
        
        return jsonify({
            "success": True,
            "data": graph_data
        })
        
    except Exception as e:
        logger.error(f"Failed to get graph data: {str(e)}\n{traceback.format_exc()}")
        return jsonify({
            "success": False,
            "error": str(e),
            "traceback": traceback.format_exc()
        }), 500


# ============== Delete Graph ==============

@graph_bp.route('/delete/<graph_id>', methods=['DELETE'])
def delete_graph(graph_id: str):
    """
    Delete Neo4j graph
    """
    try:
        if not Config.NEO4J_URI:
            return jsonify({
                "success": False,
                "error": "Neo4j connection not configured"
            }), 500
        
        builder = GraphBuilderService()
        builder.delete_graph(graph_id)
        
        return jsonify({
            "success": True,
            "message": "Graph deleted successfully"
        })
        
    except Exception as e:
        logger.error(f"Failed to delete graph: {str(e)}\n{traceback.format_exc()}")
        return jsonify({
            "success": False,
            "error": str(e)
        }), 500

File: backend-rs/src/api/graph.rs
//! Knowledge-graph endpoints (port of backend/app/api/graph.py).

use crate::error::{AppError, AppResult};
use crate::models::Ok as Payload;
use crate::services::graph::{builder, db, file_parser, ontology, projects, tasks};
use crate::state::AppState;
use axum::extract::{DefaultBodyLimit, Multipart, Path, State};
use axum::routing::{delete, get, post};
use axum::{Json, Router};
use serde::Deserialize;
use serde_json::{json, Value};

const DEFAULT_CHUNK_SIZE: usize = 500;
const DEFAULT_CHUNK_OVERLAP: usize = 50;
const MAX_UPLOAD_BYTES: usize = 50 * 1024 * 1024;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/ontology/generate", post(generate_ontology))
        .route("/build", post(build_graph))
        .route("/task/:task_id", get(get_task_status))
        .route("/data/:graph_id", get(get_graph_data))
        .route("/delete/:graph_id", delete(delete_graph))
        .route("/projects", get(list_projects))
        .route("/project/:project_id", get(get_project).delete(delete_project))
        .layer(DefaultBodyLimit::max(MAX_UPLOAD_BYTES))
}

// ============== API 1: Generate Ontology ==============

async fn generate_ontology(
    State(st): State<AppState>,
    mut multipart: Multipart,
) -> AppResult<Payload<Value>> {
    let mut files: Vec<(String, Vec<u8>)> = Vec::new();
    let mut simulation_requirement = String::new();
    let mut additional_context = String::new();
    let mut project_name = "Unnamed Project".to_string();

    while let Some(field) = multipart
        .next_field()
        .await
        .map_err(|e| AppError::BadRequest(format!("invalid multipart body: {e}")))?
    {
        let name = field.name().unwrap_or_default().to_string();
        match name.as_str() {
            "files" => {
                let filename = field.file_name().unwrap_or_default().to_string();
                if filename.is_empty() {
                    continue;
                }
                if !file_parser::is_allowed(&filename) {
                    return Err(AppError::BadRequest(format!(
                        "File type not supported: {filename}"
                    )));
                }
                let bytes = field
                    .bytes()
                    .await
                    .map_err(|e| AppError::BadRequest(format!("failed to read {filename}: {e}")))?;
                if bytes.is_empty() {
                    return Err(AppError::BadRequest(format!(
                        "File {filename} is empty or corrupted during upload"
                    )));
                }
                files.push((filename, bytes.to_vec()));
            }
            "simulation_requirement" => simulation_requirement = read_text(field).await?,
            "additional_context" => additional_context = read_text(field).await?,
            "project_name" => {
                let v = read_text(field).await?;
                if !v.trim().is_empty() {
                    project_name = v.trim().to_string();
                }
            }
            _ => {}
        }
    }

    let simulation_requirement = simulation_requirement.trim().to_string();
    if files.is_empty() {
        return Err(AppError::BadRequest("At least one file required".into()));
    }
    if simulation_requirement.is_empty() {
        return Err(AppError::BadRequest("Simulation requirement is required".into()));
    }

    // PDF parsing is CPU-bound; keep it off the async runtime.
    let document_texts: Vec<String> = tokio::task::spawn_blocking(move || {
        files
            .iter()
            .map(|(name, bytes)| {
                file_parser::extract_text(name, bytes)
                    .map_err(|e| anyhow::anyhow!("Failed to parse file {name}: {e}"))
            })
            .collect::<Result<Vec<_>, _>>()
    })
    .await
    .map_err(|e| AppError::Other(anyhow::anyhow!("file parsing task panicked: {e}")))??;

    let document_texts: Vec<String> =
        document_texts.into_iter().filter(|t| !t.trim().is_empty()).collect();
    if document_texts.is_empty() {
        return Err(AppError::BadRequest(
            "No valid text extracted from uploaded files".into(),
        ));
    }

    let additional_context = additional_context.trim().to_string();
    let mut project = projects::create(&project_name)?;
    project.simulation_requirement = Some(simulation_requirement.clone());
    project.additional_context =
        (!additional_context.is_empty()).then(|| additional_context.clone());

    let all_text = document_texts.join("\n\n");
    project.total_text_length = Some(all_text.chars().count());
    projects::save_extracted_text(&project.project_id, &all_text)?;

    let generated = ontology::generate(
        &st.llm,
        &document_texts,
        &simulation_requirement,
        (!additional_context.is_empty()).then_some(additional_context.as_str()),
    )
    .await?;

    project.ontology = Some(json!({
        "entity_types": generated["entity_types"],
        "edge_types": generated["edge_types"],
    }));
    project.analysis_summary = generated["analysis_summary"].as_str().map(String::from);
    project.status = projects::status::ONTOLOGY_GENERATED.to_string();
    projects::save(&project)?;

    Ok(Payload(json!({
        "data": {
            "project_id": project.project_id,
            "ontology": project.ontology,
            "analysis_summary": project.analysis_summary,
            "total_text_length": project.total_text_length,
        }
    })))
}

async fn read_text(field: axum::extract::multipart::Field<'_>) -> AppResult<String> {
    field
        .text()
        .await
        .map_err(|e| AppError::BadRequest(format!("invalid form field: {e}")))
}

// ============== API 2: Build Graph ==============

#[derive(Deserialize)]
struct BuildRequest {
    project_id: String,
    #[serde(default)]
    graph_name: Option<String>,
    #[serde(default)]
    chunk_size: Option<usize>,
    #[serde(default)]
    chunk_overlap: Option<usize>,
}

async fn build_graph(
    State(st): State<AppState>,
    Json(req): Json<BuildRequest>,
) -> AppResult<Payload<Value>> {
    let graph = st.graph().clone();

    let project_id = req.project_id.trim().to_string();
    if project_id.is_empty() {
        return Err(AppError::BadRequest("project_id is required".into()));
    }
    let project = projects::get(&project_id)
        .ok_or_else(|| AppError::NotFound(format!("Project does not exist: {project_id}")))?;
    let ontology = project.ontology.clone().ok_or_else(|| {
        AppError::BadRequest(
            "Project ontology not generated, please call /ontology/generate first".into(),
        )
    })?;
    let text = projects::get_extracted_text(&project_id).ok_or_else(|| {
        AppError::NotFound("Project text not found, file may have been deleted".into())
    })?;

    let graph_name = req
        .graph_name
        .filter(|n| !n.trim().is_empty())
        .unwrap_or_else(|| "Popinion Graph".to_string());

    let task_id = builder::spawn_build(
        graph,
        st.llm.clone(),
        builder::BuildParams {
            project_id,
            text,
            ontology,
            graph_name: format!("{} - {}", project.name, graph_name),
            chunk_size: req.chunk_size.unwrap_or(DEFAULT_CHUNK_SIZE).max(1),
            chunk_overlap: req.chunk_overlap.unwrap_or(DEFAULT_CHUNK_OVERLAP),
        },
    );

    Ok(Payload(json!({ "data": { "task_id": task_id } })))
}

// ============== Task Status ==============

async fn get_task_status(Path(task_id): Path<String>) -> AppResult<Payload<Value>> {
    let task = tasks::get(&task_id)
        .ok_or_else(|| AppError::NotFound(format!("Task does not exist: {task_id}")))?;
    Ok(Payload(json!({ "data": task })))
}

// ============== Graph Data ==============

async fn get_graph_data(
    State(st): State<AppState>,
    Path(graph_id): Path<String>,
) -> AppResult<Payload<Value>> {
    let data = db::get_graph_data(st.graph(), &graph_id).await?;
    Ok(Payload(json!({ "data": data })))
}

async fn delete_graph(
    State(st): State<AppState>,
    Path(graph_id): Path<String>,
) -> AppResult<Payload<Value>> {
    db::delete_graph(st.graph(), &graph_id).await?;
    Ok(Payload(json!({ "message": "Graph deleted successfully" })))
}

// ============== Projects ==============

async fn list_projects() -> AppResult<Payload<Value>> {
    let projects = projects::list(50);
    Ok(Payload(json!({
        "data": { "total": projects.len(), "projects": projects }
    })))
}

async fn get_project(Path(project_id): Path<String>) -> AppResult<Payload<Value>> {
    let project = projects::get(&project_id)
        .ok_or_else(|| AppError::NotFound(format!("Project does not exist: {project_id}")))?;
    Ok(Payload(json!({ "data": project })))
}

async fn delete_project(Path(project_id): Path<String>) -> AppResult<Payload<Value>> {
    if !projects::delete(&project_id) {
        return Err(AppError::NotFound(format!("Project does not exist: {project_id}")));
    }
    Ok(Payload(json!({ "message": "Project deleted successfully" })))
}
